######
#Qanbari Creeping Windows 
######



#Creeping Window Analyses based on Qanbari et al. (2012; PLOS ONE) 
	#http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0049525
	#This script takes in a data frame of fFst values for each position in your chromosome/scaffold
	
#Data must be in data frame format with 3 columns corresponding to chromosome/scaffold, position, and fst value IN THAT ORDER!
	#	       chr1 156 0.00070403
	#	       chr1 203 0.00008735
	#					...
	#	 	   chr4 361 0.01622014
	#	 	   chr4 410 0.00163847
	#	 	   chr4 436 0.00677428
	#		   chr4 447 0.01894072

# Edited by BAH 23-III-16



# Load Required Functions and Pakcages ------------------------
library(IRanges)
library(Hmisc)
library(ggplot2)
library(plyr)
library(reshape2)
library(RColorBrewer)
library(scales)
source("/media/data1/forty3/R/scripts/Q_correct.r")



perm.creeper <-function(n=999,  num=num, fst=creeper$fstHvL){
	#estimates perm likelihood of given SNP window generated by creeper script
	#draws SNPs within clusters to account for linkage effects
	vec=rep(0,n)
	for (i in 1:n){
		size = sample(num, 1, replace=TRUE)
		len = length(fst)-size
		samp = sample(c(1:len), 1, replace=T)
		ran = samp:(samp+size-1)
		x = mean(fst[ran])
		vec[i]=x
	}
	vec

}



# Declare Variables ----------------------------
window.size = 5000 #size of window to be crept and gaps to skip, in BP 
bin.size = 6 # minimum number of SNPs in a window (bin size, see plot)
signi = 0.01 #significance cutoff for high windows 







#with CHROMS
test = gwas
#fst.all = fst
#test$fst45 = fst.all$fst45 
#test$fst14 = fst.all$fst14 #LDB should have lower A in regions....... 
#test$fst15 = fst.all$fst15
test = test[c("CHROM","POS", "fst45", "fst14", "fst15")]
test = test[order(test$POS),]
names(test) = c("V2", "POS", "fst45", "fst14", "fst15")

			#with SCAFFs
			#test = fst 
			#test = test[c("CHROM","POS", "fst45", "fst14", "fst15")]
			#names(test) = c("V2", "POS", "fst45", "fst14", "fst15")


#Creeper Function -------------------------------

creep.all = c()
for(k in unique(test$V2)){
	
	#Gather SNPs and FST as numeric variables
	snp = as.numeric(test$POS[test$V2==k])
	fst = as.numeric(test$fst45[test$V2==k]) #fst in Hvs L
	fst2 = as.numeric(test$fst14[test$V2==k]) #fst in A vs L
	fst3 = as.numeric(test$fst15[test$V2==k]) #fst in A vs H
	
	#estimate difference between SNPs	
	p = as.numeric(test$POS[test$V2==k])
	p = p[-1]
	p = c(p, NA)
	diff = abs(as.numeric(test$POS[test$V2==k])-p)
	diff = diff[!is.na(diff)]

	x=rep(0, (max(snp)+ window.size))
	x[snp]=1
	x=cumsum(x)
	endsnp=x[snp+ window.size]
	n=length(fst)
	vec=rep(0,n)
	vec2=rep(0,n)
	vec3=rep(0,n)	
	len=rep(0,n)
	end=rep(0,n)
	cors = rep(0,n)
	cors.p = rep(0,n)
	
	for (i in 1:(n-1)){
			if (diff[i] < window.size ){
				snp_len=length(fst[i:endsnp[i]])
				
				fst_mean=mean(fst[i:endsnp[i]])
				vec[i]=fst_mean
				
				fst_mean=mean(fst2[i:endsnp[i]])
				vec2[i]=fst_mean
								
				len[i]=snp_len
				end[i]=(snp[endsnp[i]])
				
				fst_mean=mean(fst3[i:endsnp[i]])
				vec3[i]=fst_mean
				
				if(snp_len>10){
				cor = cor.test(fst[i:endsnp[i]], fst2[i:endsnp[i]])$estimate
				cors[i] = cor
				cor = cor.test(fst[i:endsnp[i]], fst2[i:endsnp[i]])$p.value
				cors.p[i] = cor
				
				}else{
				cors[i] = NA
				}
				
			}else{
				vec[i]=NA
				vec2[i]=NA
				vec3[i]=NA
				cors[i] = NA
				cors.p = NA
				len[i]=NA
				end[i]=(snp[endsnp[i]])
				
			}

}


	creeper = data.frame(cbind(vec,vec2,vec3, len, snp, end, cors, cors.p))
	creeper$CHROM = rep(k, nrow(creeper))
	names(creeper)=c("fstHvL","fstLvA", "fstHvA","num_snps", "start", "end", "cor", "p","chrom")
	creep.all = rbind(creep.all, creeper)



}
	
creep.all = creep.all[which(complete.cases(creep.all$num_snps)),]
creep.all = creep.all[which(creep.all$num_snps>0),]


# Trim windows with too few SNPs -----------------------------
	# mean FST is going to be biased in small windows
	#This lets you find a cut-off for minimum window size
	#I've not yet automated this, but choose the bin that has consistent variation
	#	default is bin 10, if not changed 

len = round(max(creep.all$num_snps)/10) 
creep.all$bin <- as.numeric(cut2(creep.all$num_snps, g=len))
var.plot = aggregate(creep.all$fstHvL, by = list(creep.all$bin), sd)


pdf(file="NumSNPsbyBin.pdf")
	plot(creep.all$bin, creep.all$fstHvL)
dev.off()


pdf(file="SDbyBin.pdf")
	plot(var.plot,pch=16)
dev.off()


creeper = creep.all
creeper=creeper[(creeper$bin > bin.size),]



# Estimate differential admixture -------------------------------
	#As per Zayed and Whitfield (2008; PNAS)
	#FstHvA/(FstHvA+FstHvL) which ranges from 0 to 1, indicating no to complete introgression of non-A alleles into aggressive bees. 

creeper$HvAint = creeper$fstHvA / (creeper$fstHvA + creeper$fstLvA)



#Significance Testing ---------------------------------------------
	# Permutation test
		#selects random window size from distribution of window sizes in data (W)
		# selects W SNPs beside each other
		#estimates mean FST of those random SNPs
		#repeats N times 
	#corrects for multiple testing with Q test (Storey Method

	
num = creeper$num_snps
lec=perm.creeper(n=100000, num=num)

	#hist(lec,breaks=100)
	#x11();hist(creeper$fstHvL, breaks=100)
creeper$pnorm.fst=pnorm(creeper$fstHvL, mean=mean(lec), sd=sd(lec),lower.tail=F)
#plot(pnorm.fst, creeper$fstHvL)
creeper$q=qvalue1(creeper$pnorm.fst)$q



save.image("creeperAFZ.RData")
	
	#now, overlap with QTLs, genes, etc.
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
#Assemble overlapping, significant windows ------------------------------------
high = creeper[which(creeper$q<signi ),] #or signi/1000
high$group2 = rep("NA", nrow(high))

for(k in unique(high$chrom)){
	ir <- IRanges(high$start[which(high$chrom==k)], high$end[which(high$chrom==k)]) 
	high$group2[which(high$chrom==k)] <- subjectHits(findOverlaps(ir, reduce(ir)))
}


rang=aggregate(start~chrom + group2,data = high, min)
rang.1=aggregate(end~chrom + group2, data = high, function(x) max(x))
rang$end = rang.1$end


write.table(rang[c(1,3,4)],file="HighFSTWindows", col.names=F, row.names=F,quote=F)






#PLOTTING









sig = unlist(creeper$q)
sig[which(creeper$q<0.05 & creeper$p<0.05)] = "pink"
sig[which(creeper$q<0.05 & creeper$p>0.05)] = "red"
sig[which(creeper$q<0.001)] = "red"
#sig[which(sig!="red" & sig!="pink")] = "black"	
sig[which(sig!="red")] = "black"	
sig[which(is.na(sig))] = "black"



#plot(creeper$start[creeper$chrom==chr],creeper$HvAint[creeper$chrom==chr],col=sig[creeper$chrom==chr], pch=19)
x11();plot(creeper$start[which(creeper$chrom==chr)],creeper$fstHvL[which(creeper$chrom==chr)],col=sig[which(creeper$chrom==chr)], pch=19)





 

df.fst = creeper
sig = unlist(creeper$q)
sig[which(creeper$q<signi )] = "orange"
sig[which(creeper$q<(signi/1000))] = "red"

#sig[which(creeper$q<0.01 & creeper$p<0.05)] = "blue"
#sig[which(creeper$q<0.001)] = "red"
sig[which(sig!="red" & sig!="orange")] = "black"	
#sig[which(sig!="red" & )] = "black"	
sig[which(is.na(sig))] = "black"

par(mfrow = c(8, 2),
	cex = 0.4, 
	mar = c(1, 2, 2, 2), 
	oma = c(1, 1, 1, 1)
	)
for(i in 1:16){

	plot(as.numeric(unlist(df.fst$start[df.fst$chrom==i])),as.numeric(unlist(df.fst$fstHvL[df.fst$chrom==i])), 
		pch=19,
		cex=0.3,
		col = c(sig[df.fst$chrom==i] ),
		axes=FALSE,
		xlab=NA,
		ylab=NA,
		xlim=c(0,  30000000),
		ylim=c(0,0.025)
	)
	xpos <- seq(0, max(as.numeric(unlist(df.fst$start[df.fst$chrom==i]))), by=1000000)
	axis(1, at=xpos,labels=xpos/1000)
	#axis(2, pos=0, at=c(0,4,8,12), labels=c("","4","8","12"))
	

#add in QTLs	
	
	if(i==12){
		rect(xleft = 9035719, ybottom=-0.05, xright=11902654, ytop=12, 
		border = NA,
		col=adjustcolor( "grey", alpha.f = 0.2) 
		)
		rect(xleft = 7988527, ybottom=-0.05, xright=8985718, ytop=12, 
		border = NA,
		col=adjustcolor("grey", alpha.f = 0.2)
		)
	}else{
	if(i==3){
		rect(xleft = 1393932, ybottom=-0.05, xright=2280481, ytop=12, 
		border = NA,
		col=adjustcolor( "black", alpha.f = 0.2) #oxley
		)
	}else{
	if(i==7){
		rect(xleft = 1583221, ybottom=-0.05, xright=1935038, ytop=12, 
		border = NA,
		col=adjustcolor( "black", alpha.f = 0.2) #oxley
		)
		rect(xleft = 1985039, ybottom=-0.05, xright=2268329, ytop=12, 
		border = NA,
		col=adjustcolor( "black", alpha.f = 0.2) #oxley
		)		
		
		
	}else{		
	
	}
	}
	}
mtext(i,side = 3, line = -1, adj = 0.05, cex = 1)
}










































########
#comparing AvL and LvH
	#rang = AvL
	#rang45 = LvH
rang$group2=NULL
rang45 = read.table(file="HighFSTWindows",header=F); names(rang45)=names(rang)
rang45$comp = rep("HvL", nrow(rang45))
Genic.Fst = c()
for(i in intersect(rang$chrom, rang45$chrom)){
	win.temp = rang[rang$chrom==i,]
	deg.temp = rang45[which(as.character(rang45$chrom)==as.character(i)),]
	blah=outer(as.numeric(deg.temp$start-1000), as.numeric(as.character(win.temp$end)), "<=") 
	blah1=outer(as.numeric(deg.temp$end+1000), as.numeric(as.character(win.temp$start)), ">=") 
	blah=(which(blah1=="TRUE" & blah=="TRUE", arr.ind=T)) #The gene region will be the colum variable
	temp = deg.temp[blah[,1],]
	temp = cbind(temp, win.temp[blah[,2],])
	Genic.Fst = rbind(temp,Genic.Fst)
	print(i)
}


	
	



###########




    chrom    start      end comp chrom    start      end
275  chr9  8197571  8202571  HvL  chr9  8198596  8203584
89   chr7 13148403 13158126  HvL  chr7 13144862 13153404
219  chr6 13975662 13983516  HvL  chr6 13981855 13986843
251  chr6  3175776  3185424  HvL  chr6  3172026  3186377
272  chr6  3204490  3214286  HvL  chr6  3200331  3205821
288  chr6  3434494  3439486  HvL  chr6  3433201  3440293
130  chr4   952449   957103  HvL  chr4   954069   963642
76   chr4  9563136  9567586  HvL  chr4  9563136  9567586
138  chr2  8192258  8207108  HvL  chr2  8190954  8199603
7   chr15   447307   457257  HvL chr15   452420   462151
196 chr15   530900   537882  HvL chr15   534770   539627
330 chr15  4245683  4255627  HvL chr15  4247517  4256509
6   chr14  1227124  1234371  HvL chr14  1231459  1236741
294 chr11 10612824 10617384  HvL chr11 10616450 10621695
19  chr11 11911728 11919416  HvL chr11 11911322 11926088
33  chr11 11919782 11925404  HvL chr11 11911322 11926088
47  chr11 14250424 14255422  HvL chr11 14247822 14261247
240 chr10  3481729  3486581  HvL chr10  3482979  3494231
160  chr1 24267600 24272596  HvL  chr1 24262485 24267653
175  chr1 27877108 27883890  HvL  chr1 27877122 27883192
180  chr1 27889915 27904335  HvL  chr1 27888878 27895202



# Plot the results -------------------------------------------------------------


#plot data	
#pdf(file="ApisBombusGammaHistogram.pdf",width=16,height=10)

creep.plot = creep.all[creep.all$bin > bin.size,]
#creep.plot = creep.plot[creep.plot$chrom %in% c("chr12","chr6","chr10"),]
creep.plot$chrom = factor(creep.plot$chrom, 
	levels = c("chr1","chr2","chr3",
	"chr4","chr5","chr6",
	"chr7","chr8","chr9",
	"chr10","chr11","chr12",
	"chr13","chr14","chr15",
	"chr16"	))
max.fst = round(range(high$fst)[1],3)


	
p<-ggplot(creep.plot, aes(x= start, y = fst))+
	geom_point(color="lightgrey", shape=19, size = 0.001)+
	coord_cartesian(ylim = c(0.004, 0.02)) + 
 	
	geom_hline(yintercept = max.fst , lty = 2) +
	
	facet_grid(chrom ~ .) +
	
	theme_classic() + 
	
	theme(
		strip.text.y = element_text(size = 12),
		axis.title.x = element_text(size = 12),
		axis.text=element_text(size=10)	,
		axis.line.x = element_line(colour = 'black', size=0.5, linetype='solid'),
		axis.line.y = element_line(colour = 'black', size=0.5, linetype='solid'))+
	scale_x_continuous(labels = comma) +
	labs(
		y= "Fst",
		x = "Position (bp)")


		
		






# -------------------
gff = read.table(file="/media/data1/forty3/brock/scripts/GFF/NCBIGFF.txt")
gff$V1 = paste("chr", gff$V1, sep="")
names(gff)[1]="chrom"
gff = gff[gff$V5=="gene",]

Genic.Fst = c()
for(i in intersect(gff$chrom, rang$chrom)){
	win.temp = gff[gff$chrom==i,]
	deg.temp = rang[which(as.character(rang$chrom)==as.character(i)),]
	
	blah=outer(as.numeric(deg.temp$start), as.numeric(as.character(win.temp$V3)), ">=") 
	blah1=outer(as.numeric(deg.temp$end), as.numeric(as.character(win.temp$V4)), "<=") 
	blah=(which(blah1=="TRUE" & blah=="TRUE", arr.ind=T)) #The gene region will be the colum variable
	
	temp = deg.temp[blah[,1],]
	temp = cbind(temp, win.temp[blah[,2],])
	Genic.Fst = rbind(temp,Genic.Fst)
	print(i)


}








# Other --------------------------









